New user button to sidebar

skjnldsv opened the pull request:
Fix #6746
Fix #7162 
@nextcloud/designers 

![capture d ecran_2018-02-23_12-05-25](https://user-images.githubusercontent.com/14975046/36591303-da2b5c64-1891-11e8-9935-a16a60de3cff.png)

- [x] Do we need the confirm submit button?
- [x] On mobile, clicking the new button hides the sidebar
- [x] Update documentation

skjnldsv committed:
New button standard

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

USer menu in popover

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

Multiselect fixes

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

Register menu & copyright

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

Allow form and label in popover standard

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

New menu NOT in popover

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

Stop autofilling user and password...
SHAME TO THEM: https://bugzilla.mozilla.org/show_bug.cgi?id=956906#c100

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

Hide men by default

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

juliushaertl commented:
I'm not that sure if we should go with a popover here. I'd rather use the solution @jancborchardt proposed in #7162:

> That would add a new row in the content area on top of the table, with empty fields and dropdowns using default values, with the username field focused.

Makes much more sense to enter the new user data inline where it will appear later on as well.

pixelipo commented:
I agree with @juliushaertl - also because @jancborchardt 's proposed solution would also solve #6468

skjnldsv commented:
Okay :)

MariusBluem commented:
Even if I am the only one maybe, I like the proposed change.

1) moving the „Add user“ button into the sidebar makes sense, since Mail and Contacts have the button at the same place. It makes the interface more clean and - at least for me - more easy to use. I would even love to see this in the Files UI.

2) we can discuss the design of the appearing dropdown, there may be place for polishing.

pixelipo commented:
Nobody is contesting moving the Add button to the sidebar - that is a good solution. It's just that the button shouldn't open a popover, by rather create a new row with input fields in the main container.

It's a UX general rule of thumb that models should be used rarely.

In this particular example it's easy to create a usability issue - lose focus from popover and what happens with data that was already input?

MariusBluem commented:
Ah ... okay, sorry for the misunderstanding :)

MariusBluem commented:
maybe we should remove the "+add group" button by doing this ... looks strange having to such buttons in the sidebar (with one more prominent than the other) .. "+ add group" is not needed since a group can be created by simply adding an non-existing group to a user. What do you think @nextcloud/designers

skjnldsv commented:
I agree with @MariusBluem

jancborchardt commented:
Let's keep separate things separate. :)

This issue is about redoing the "Add new user" flow. This includes adding a button in the sidebar (good start @skjnldsv!), which then adds a new row with inputs at the top of the list like I described and as @juliushaertl referemced above.

What we do with the "Add group" button is separate from this and should not block it.

skjnldsv commented:
Okay, pr updated, left to do:
- [ ] Dropdown select off by 2 px 
![capture d ecran_2018-01-08_05-50-25](https://user-images.githubusercontent.com/14975046/34659324-e784efa2-f437-11e7-8e8b-c1f417715186.png)

jancborchardt commented:
@skjnldsv sorry for the confusion, but the placement is still off. ;)

> That would add a new row **in the content area on top of the table** […]

So basically look like a new row already. Basically exactly like it looks when you **edit** a user.

That way it's symmetric for people to use, and we don't have to maintain 2 separate interfaces. :)

skjnldsv commented:
Ahahaha I'm a fool :p 
Apparently I forgot how to read 😆

skjnldsv committed:
New user form in table

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

Added value and empty check to properly display a confirm button/input

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

w3c html form table compliance and menu fixes

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

Various design fixes

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

Fix wording to consistent 'Add user'

Signed-off-by: Jan-Christoph Borchardt <hey@jancborchardt.net>

Focus new username input on toggle

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

Removed  unwanted th after rebase

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

quote fix

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

Th to td

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

🙈

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

Email input to email type

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

Fixed table template cells and fix email input enabling

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

Always show email and fixed min-width of name, username, mail and fullname columns

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

Use button id

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
Okay, what do you think now? @jancborchardt

jancborchardt commented:
Looks great! :) Just change the button label from "Create" to "Add user" (as we never use "Create") and it's good.

Please also check @nextcloud/designers :)

MariusBluem commented:
![b20f7165-235d-47d2-950f-75867f341d73](https://user-images.githubusercontent.com/19711361/34743952-e6067bdc-f58b-11e7-986a-9c0e65da74b4.jpeg)

Why dont you put the „Add User“ arrow in a line together with the 3-dot-menu. This would make room for additional settings like Group Admin, Quota and maybe E-Mail (according to what is shown/enabled) already while adding the user. What do you think? @jancborchardt

skjnldsv commented:
> Why dont you put the „Add User“ arrow in a line together with the 3-dot-menu. This would make room for additional settings like Group Admin, Quota and maybe E-Mail (according to what is shown/enabled) already while adding the user. What do you think? @jancborchardt

I did at first, but we have multiple rows hidden. So if the user choose to display all of them (in the settings) the create button goes outside the window :(

jancborchardt commented:
@MariusBluem @skjnldsv yeah, I also saw it but came to the conclusion it’s fine. :) In the future we can improve it and show whatever people display, but this is fine for now.

jancborchardt reviewed (dismissed):
@skjnldsv could you make sure the first field »Username« is automatically focused when the button is clicked and row is created?

jancborchardt commented:
I fixed the wording to "Add user". :)

skjnldsv commented:
@nextcloud/designers please review :)

skjnldsv commented:
Rebased with latest #7790 by @juliushaertl

danxuliu commented:
Minor detail:
-  Click _Add user_
-  Click outside the new user row -> the new user row is hidden; no problem here

But:
-  Click _Add user_
-  On the new user row, click _Groups_; a menu is shown
-  Click outside the menu
    -  If you click outside the menu but on the new user row -> the menu is closed
    -  If you click outside both the menu and the new user row -> the new user row is hidden

In the last case the menu should be closed, but not the new user row; of course, once the menu is closed, clicking outside the new user row should close it like above.

skjnldsv commented:
@danxuliu I'm using the menu function integrated into core. I need to think of this. Not sure if possible without rewritting a lot of stuff! :)

MorrisJobke commented:
I rebased this one and it looks really nice 👍

skjnldsv commented:
Thanks @MorrisJobke ! :)

MorrisJobke commented:
> In the last case the menu should be closed, but not the new user row; of course, once the menu is closed, clicking outside the new user row should close it like above.

Clicking again "Add user" the typed in stuff is till available. So this is good and bad, but I would say we should refine in an upcoming PR. Also if you select "send email to new user" but don't select "show email" it is not possible to create the user because an error is shown. We could improve but it works as of now 👍

MorrisJobke commented:
Time for review I would say.

juliushaertl commented:
Looks good to me, only what @MariusBluem mentioned is still an issue to me: https://github.com/nextcloud/server/pull/7600#issuecomment-356419566

The Submit button should be at the end of the line, ideally just the arrow icon without the add user text.

skjnldsv commented:
@juliushaertl like I said:
> I did at first, but we have multiple ~~rows~~ columns hidden. So if the user choose to display all of them (in the settings) the create button goes outside the window :(

I really don't have an idea on this :/

MorrisJobke commented:
> The Submit button should be at the end of the line, ideally just the arrow icon without the add user text.

Let's leave it there for now and then fix this iteratively.

juliushaertl commented on settings/templates/users/part.userlist.php:
```diff
@@ -1,90 +1,133 @@
+				<th class="name" scope="row"></th>
+				<td class="displayName"><span></span> <img class="action"
+					src="<?php p(image_path('core', 'actions/rename.svg'))?>"
+					alt="<?php p($l->t("change full name"))?>" title="<?php p($l->t("change full name"))?>"/>
```
Can you use single quotes here?

juliushaertl commented on settings/templates/users/part.userlist.php:
```diff
@@ -1,90 +1,133 @@
+				</td>
+				<td class="password"><span>●●●●●●●</span> <img class="action"
+					src="<?php print_unescaped(image_path('core', 'actions/rename.svg'))?>"
+					alt="<?php p($l->t("set new password"))?>" title="<?php p($l->t("set new password"))?>"/>
```
also single quotes :wink:

pixelipo commented on settings/templates/users/part.userlist.php:
```diff
@@ -1,90 +1,133 @@
+			<!-- the following <tr> is used as a template for the JS part -->
+			<tr style="display:none">
+				<td class="avatar"><div class="avatardiv"></div></td>
+				<th class="name" scope="row"></th>
```
Why is this a header `<th>`? It's not the title of the row/column

pixelipo commented on settings/templates/users/part.userlist.php:
```diff
@@ -1,90 +1,133 @@
+				<th class="lastLogin" scope="col"><?php p($l->t('Last login')); ?></th>
+				<th class="userActions"></th>
+			</tr>
+			<tr id="newuserHeader" style="display:none">
```
Definitely not a header - it's a regular row, so all the elements should be `<td>`

pixelipo replied:
Shouldn't this be inside the `<tbody>` instead?

skjnldsv replied:
I asked myself the same question, but I figured it should be separated from the static values, but thinking again, I think it's not such great coding to let it in the header 🤔 
I'll move it in the body! :)

pixelipo replied:
The issue is semantical, because header is reserved for:
> The table head and table foot should contain information about the table's columns. The table body should contain rows of table data.
https://www.w3.org/TR/html401/struct/tables.html#edef-THEAD

skjnldsv replied:
OOh, I remember why I choose that. Because the entire body is being replaced on load (xhr user data fetched). I placed it here to avoid this.

pixelipo commented:
this means you need to add 
```
<?php if(is_array($_['subadmins']) || $_['subadmins']): ?>
	<td></td>
<?php endif;?>
```
to `#newuserHeader`

as well as:
```
<?php if((bool)$_['recoveryAdminEnabled']): ?>
	<td></td>
<?php endif; ?>
```
to the **user** rows

and a similar:
```

```
<?php if((bool)$_['recoveryAdminEnabled']): ?>
	<th id="recoveryPassword" scope="col"><?php p($l->t('Recovery password')); ?></th>
<?php endif; ?>
```
to the header

skjnldsv commented:
@pixelipo Except enabling/disabling columns are also done in js iirc, which means I also need to add stuff in there.

Also, it will look very strange with empty cells. The best would have been to calculate how many empty cells there is and merge them to have only one big going all over to the end :)

pixelipo commented:
@skjnldsv  it doesn't look strange with empty cells - they are invisible anyway:
![image](https://user-images.githubusercontent.com/4129927/36254716-529d7156-124c-11e8-94e5-fb2b6d3fd5d2.png)
Though this does make the "Add User" button less pronounced.

Also, I checked JS on this table and I could only find instances of class/visibility toggling, no actual node manipulation, so it shouldn't be a real problem.

skjnldsv commented:
@pixelipo check the settings options :)

pixelipo commented:
@show me the actual code where that happens, @skjnldsv

skjnldsv commented:
https://github.com/nextcloud/server/blob/master/settings/js/users/users.js#L1157

MorrisJobke commented:
Conflicts and one integration test fails:

```
 And I create user unknownUser with password 123456acb             # UsersSettingsContext::iCreateUserWithPassword()
      │ User name field for new user in Users Settings value could not be set
      │ Exception message: Element is not currently visible and so may not be interacted with
```

Form the `login.feature` - cc @danxuliu

pixelipo commented:
@skjnldsv the above doesn't affect DOM, only toggles CSS.

MorrisJobke commented:
@skjnldsv Github tells me that there are still conflict 😉

skjnldsv commented:
@MorrisJobke Fixed, it was because of #8280

pixelipo commented:
Come on, @skjnldsv are you kidding me? In the code you are showing above, only this line is relevant to our discussion:
`$("#userlist .userBackend").show(); `
It says: "find the element with a class `userBackend` and modify its css from `display:none` to `display: table-cell`".
Element in question **exists** at all times - both in DOM and in the PHP template.

Really, fixing this requires no Javascript whatsoever.

skjnldsv commented:
@pixelipo 
> as well as:
```
<?php if((bool)$_['recoveryAdminEnabled']): ?>
	<td></td>
<?php endif; ?>
to the user rows
```

This will create empty fiels all over the table.
I agree with you , it would be better visually, but it's basically a mess we're introducing. Having empty cells poping here and there for the purpose of just aligning a single button doe snot seems like the proper thing to do. sorry :/

skjnldsv commented:
> Come on, @skjnldsv are you kidding me? In the code you are showing above, only this line is relevant to our discussion:

No need to be rude please! 😕 
Because it will display cells in some rows and therefore break the table layout by having a disparate number of cells across the table :)

pixelipo commented:
This PR adds random cells to a table.

My proposal is actually less messy, because it adds entire row.

Table is a grid. Each Each column should have the same amount of cells. Sage goes for rows.
Cells can be merged, but that's not for elegance, but special use case when you have sub-headers or multiple items in a single row-column tuple

pixelipo commented:
I meant adds entire column

MorrisJobke commented:
> I agree with you , it would be better visually, but it's basically a mess we're introducing. Having empty cells poping here and there for the purpose of just aligning a single button doe snot seems like the proper thing to do. sorry :/

I would be fine with this. Then also the cleanup of this "mess" would be easier, because you clean up the same amount of cells in all implementations. 🤷‍♂️

skjnldsv commented:
Okay, one step forward, we still have one issue tough. The `show email address` and `send email confirmation` both uses the mail column.
If we disable the mail for all user but then click the "send email confirmation", we get this:
![capture d ecran_2018-02-15_13-29-51](https://user-images.githubusercontent.com/14975046/36256749-81aec532-1254-11e8-9a05-b263420943de.png)


1. we bind the two options together and show the email column for every rows if enabled
2. ???, Any idea?

MorrisJobke commented:
> we bind the two options together and show the email column for every rows if enabled

At least in one way: if "send email to users" the "show email" is automatically enabled as well and if "show email" is disabled the "send email to users" is disabled as well. The other way around should work nevertheless. (just a little JS should be fine here - whoever tricks the JS will get the error from the server anyways)

pixelipo commented on settings/templates/users/part.userlist.php:
```diff
@@ -1,90 +1,146 @@
-					<?php foreach($_['quota_preset'] as $preset):?>
-						<option value='<?php p($preset);?>'>
-							<?php p($preset);?>
+<form class="newUserMenu" id="newuser" autocomplete="off">
```
Perhaps a stupid question, but shouldn't this be wrapped only around `#newuserHeader`? Rest of the table is not part of the form, is it?

skjnldsv replied:
There is no stupid questions :)

I had an issue about this, and form is not allowed inside a table, so either we have to create 2 tables, or wrap the whole table in the form. Do you see an inconvenience?

pixelipo replied:
Answers my question :)

MorrisJobke commented:
Was discussed in IRC: it just makes most sense to always show the email and not mess around too hard with those two settings.

pixelipo commented on settings/js/users/users.js:
```diff
@@ -31,6 +32,13 @@ var UserList = {
 		// initialize these entries
 		this.$el.find('.quota-user').singleSelect().on('change', this.onQuotaSelect);
+		OC.registerMenu(
+			$('.app-navigation-new > button'),
```
I would feel safer if this was using id of the element. ID's are unique, classes can be anywhere. In addition, `.app-navigation-new > button` could be used all over the place.
Also, why is the button itself not the target? Mail app, for example is using button's ID:
`<button type="button" id="mail_new_message" class="icon-add">New message</button>`

skjnldsv replied:
Done :)

pixelipo commented on settings/css/settings.scss:
```diff
@@ -491,9 +491,12 @@ table.grid {
 	&.name {
 		padding-left: .8em;
+		width: 10em;
+		min-width: 10em;
```
Any reason for adding min and max widht when width has already been set?

pixelipo commented on settings/css/settings.scss:
```diff
@@ -504,8 +507,12 @@ td {
+	&.mailAddress {
 		width: 12em;
+		min-width: 12em;
+		max-width: 12em;
```
Any reason for adding min and max widht when width has already been set?

skjnldsv replied:
because it's being ignored in the tables, this is a known issue for years. :/

skjnldsv replied:
Do you suggest `table-layout:fixed` ?

pixelipo replied:
I'm not suggesting anything, let's keep that ugly hack for now :)

pixelipo commented:
@skjnldsv I don't think I was rude - I didn't insult you in any way. I just really hate having to repeat myself.

Regarding the messiness, table must have same amount of cells in each row/column. Merged cells don't count as a single cell.
You can't have a valid HTML table in which some rows have 5 columns and some have 7 columns.
Table is "X times Y" cells.

So, I'm glad you were able to implement that without messing about with JS ;)

skjnldsv commented:
> I didn't insult you in any way. I just really hate having to repeat myself.

You can be rude without insulting people. And we're on a big project where a lot of people are volunteer, like me. If you don't like to repeat yourself, then don't, I did not asked you to and I do not work for you either. We're all trying to push this project towards a better place.

skjnldsv commented:
Okay, rebased!
@pixelipo could you update your review please? :)

pixelipo reviewed (dismissed):
I think two things still need fixing:
1. `#newuserHeader` row is hidden when it loses focus. It retains content, but it's a bit weird.
2. "Add user" button - it's too large and messes up the layout. Checkmark icon in Deck app looks quite decent - maybe we should try that?
![image](https://user-images.githubusercontent.com/4129927/36584961-2a7f27e2-187c-11e8-8361-6f034190acb4.png)

skjnldsv committed:
Changed button types and do not hide form on outside click

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

pixelipo commented:
@skjnldsv mush better in my opinion, but perhaps it should look more like this?
![image](https://user-images.githubusercontent.com/4129927/36599335-5fa981d2-18af-11e8-8cc3-ea0c84212d36.png)

```
#newuserHeader .userActions input {
    border: none;
    background-color: initial;
    opacity: .5;
    width: 44px;
    height: 44px;
}

skjnldsv commented:
Makes more sense to unify the look with the menu button under it yeah! :)

skjnldsv committed:
Unify input looks

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
@pixelipo good to go :)

jancborchardt commented:
@skjnldsv @MorrisJobke as discussed before, we can completely get rid of both settings pertaining to the email, and simply show the email column always, as well as send the notification mail always (as long as email is set). I thought that to be separate from this, but if it ties in so much maybe best to fix it here.

@pixelipo it's not good to have the destructive action being so close to the confirm action, especially when they are both only icons.


And @pixelipo @skjnldsv relax :) text communication is sometimes difficult and stuff needs to be repeated. Repeating is basically what I've been doing most. :D

skjnldsv commented:
@jancborchardt so do you want me to add the text?

pixelipo commented:
@jancborchardt I don't really see a good solution to the closeness of those 2 icons. 

I think "Cancel" is not a totally destructive action, since it doesn't remove existing data, only the non-submitted form data. Previously focus-out was used to hide the form itself, which is not a good solution either.
I think this look is the only one that doesn't break UI, so it makes no sense to change the CSS of the actions.
Removing the "Cancel" button will make it hard to hide the form once it's been revealed.

Do you have any suggestion?

skjnldsv commented:
Agree with the primary! @pixelipo ? :)

rullzer reviewed (approved):
Works as advertised!

skjnldsv committed:
Primary to confirm button

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
![capture d ecran_2018-02-27_13-11-11](https://user-images.githubusercontent.com/14975046/36728043-b678a11a-1bbf-11e8-9498-0c1c4101ed25.png)

skjnldsv commented:
Wait a bit before merging please @rullzer

MorrisJobke commented:
Acceptance tests fail:

```
cenario: log in with invalid user once fixed by admin              # /drone/src/github.com/nextcloud/server/tests/acceptance/features/login.feature:32
    Given I act as John                                               # ActorContext::iActAs()
    And I can not log in with user unknownUser and password 123456acb # LoginPageContext::iCanNotLogInWithUserAndPassword()
    When I act as Jane                                                # ActorContext::iActAs()
    And I am logged in as the admin                                   # LoginPageContext::iAmLoggedInAsTheAdmin()
    And I open the User settings                                      # SettingsMenuContext::iOpenTheUserSettings()
    And I create user unknownUser with password 123456acb             # UsersSettingsContext::iCreateUserWithPassword()
      │ User name field for new user in Users Settings value could not be set
      │ Exception message: Element is not currently visible and so may not be interacted with
      │ Build info: version: '2.53.1', revision: 'a36b8b1', time: '2016-06-30 17:37:03'
      │ System info: host: '47d48347a2f2', ip: '172.17.0.13', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-96-generic', java.version: '1.8.0_91'
      │ Driver info: driver.version: unknown
      │ Trying again
      │
      Element is not currently visible and so may not be interacted with
```

skjnldsv committed:
Fix acceptance tests

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
>     And I set the password for user0 to 654321                               # UsersSettingsContext::iSetThePasswordForUserTo()
>       Offset within element cannot be scrolled into view: (0, 0): [object HTMLTableCellElement]
>       Build info: version: '2.53.1', revision: 'a36b8b1', time: '2016-06-30 17:37:03'
>       System info: host: 'a18942a51f1e', ip: '172.17.0.17', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-96-generic', java.version: '1.8.0_91'
>       Driver info: driver.version: unknown (WebDriver\Exception\MoveTargetOutOfBounds)

Fixed most of it, will see tomorrow if danxuliu did not picked this up before me :)

skjnldsv committed:
Sticky header in users list

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
fixup! Sticky header in users list

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

MariusBluem commented:
Will you integrate https://github.com/nextcloud/bookmarks/pull/424#issuecomment-368908888 in this PR?

danxuliu commented:
The acceptance tests should all work now; I have squashed all the fixes related to the acceptance tests in a single commit.

skjnldsv commented:
Failure unrelated.
Acceptance now passes
